### PR TITLE
v10.0.0 release

### PR DESCRIPTION
**Pull request by wernerfred**
# Description

This PR marks the feature freeze phase for `v10.0.0` release. The current `:edge` image will be tested for roughly one week and if no bugs get reported `v10.0.0` will be released. During this testing period no new PRs will be merged.

**If you want to support us give the latest `:edge` image a try and report any issues you might (hopefully not) encounter back to this PR.**

## Type of change

- [x] Release

## Checklist:

- [ ] Update `CHANGELOG.md`
- [x] Update references in `README.md`
- [x] Update version in `VERSION`
- [x] ~~Merge #1921~~
- [x] Merge #1951
- [x] Merge #1989 
- [x] ~~Merge #1996~~ (closed)
- [x] Merge #2021

### CODE REVIEW

**Comment by wernerfred** [comment]
I created a milestone (we should use this in future to save us time looking for all changes since last release) for `v10.0.0` and linked all PRs since `v9.1.0` that needs to be included in the `CHANGELOG` (i left out PR fixes and dep-bumps). Feel free to double check!

https://github.com/docker-mailserver/docker-mailserver/milestone/7

**Comment by polarathene** [comment]
> we should use this in future to save us time looking for all changes since last release

Isn't that the purpose of conventional commits? (although it doesn't seem like that has been consistent in commit naming on master)

There is tooling that can manage the changelog that way instead of manually handling it. I haven't looked at the commit history though, so if it's not that much of a hassle probably no need to bring another tool into the mix for that :sweat_smile:

**Comment by aendeavor** [comment]
I would not yet introduce more tooling. This projects still evolves very rapidly ATM and we're busy enough :D

**Comment by wernerfred** [comment]
> Isn't that the purpose of conventional commits?

Yes but that presumes that everyone uses conventional commit messages or they are enforced. This would definitely be a bigger amount of work instead of linking PRs to milestones for now (someone needs to press merge anyways so manual action from one of us is involved every time).

I woul like to see a fully automated approach in the future but i am not even sure if conventional commits is the right approach (for all?) although it is a very good one. For now: 

> I would not yet introduce more tooling.

**Comment by wernerfred** [comment]
@aendeavor i removed #1921 from the top list as this is only a doc update and therefore unrelated to a docker image version. It can be merged anytime without impact on the build or release process

**Comment by aendeavor** [comment]
> @aendeavor i removed #1921 from the top list as this is only a doc update and therefore unrelated to a docker image version. It can be merged anytime without impact on the build or release process

I wanted this to be in v10 so users would see the improved doc when viewing the docs for v10. But it's fine with me if the is not on v10 after all.

**Comment by wernerfred** [comment]
> I wanted this to be in v10 so users would see the improved doc when viewing the docs for v10. But it's fine with me if the is not on v10 after all.

If @moqmar finds time to finish his work i have no objection to merge it before releasing `v10.0.0`
As we plan to test at least one week or two he has time until close to release/end of test phase.

**Comment by aendeavor** [comment]
> > I wanted this to be in v10 so users would see the improved doc when viewing the docs for v10. But it's fine with me if the is not on v10 after all.
> 
> 
> 
> If @moqmar finds time to finish his work i have no objection to merge it before releasing `v10.0.0`
> 
> As we plan to test at least one week or two he has time until close to release/end of test phase.

He said he was going to do it tomorrow or in the next few days. Let's see. If he finished in time, we can merge it. There is however the issue with the LDAP variable not being set correctly. I think this should be merged before v10. What's your take?

**Comment by wernerfred** [comment]
> What's your take?

If finished in time we should merge it :)

**Comment by casperklein** [comment]
As this is breaking change, I would also include it in v10. Otherwise this would addressed at the earliest in v11.

**Comment by aendeavor** [comment]
@wernerfred you added #1996 with 58a8ba4b440aab637e2dc7df1856d6fef08ebd46, but this PR was closed without being merged:)

**Comment by wernerfred** [comment]
As we are testing `:edge` I added help wanted on purpose but one can argue about it^^ 
No big deal :D

**Comment by aendeavor** [comment]
The feature freeze is now in place for over seven days without major incidents. This PR is ready for final review. @docker-mailserver/maintainers Please have a look!

**Comment by wernerfred** [comment]
> @wernerfred is the CHANGELOG up to date? The tick is still missing in the description.

CHANGELOG schould be up to date. Tick is missing bc there were/might be fix PRs missing and we merged a lot during this phase so i wanted to make sure that we keep an eye on it. 
Keep in mind that the Changelog notice in the v10 release draft it heavily out of date and needs to be replaced with htis one after merging.

> The feature freeze is now in place for over seven days without major incidents. 

Had someone the chance to test a LDAP environment? My setup without LDAP is running fine for ~ a week, hadn't time to test the LDAP setup as well.

File: docs/content/config/advanced/optional-config.md
Comment (polarathene) [inline docs/content/config/advanced/optional-config.md:14]: Dropping `Auto-configurable via [`setup.sh config ssl`][docs-setupsh].` okay. Is this directory no longer relevant to document?

Is there no way this directory is generated/used now?

The referenced doc page does have some content referencing it [and the command that is being removed](https://docker-mailserver.github.io/docker-mailserver/edge/config/security/ssl/#self-signed-certificates). While the command may not be relevant, the section still is. Note that I have documented the generation of our test certs [here](https://github.com/docker-mailserver/docker-mailserver/tree/master/test/test-files/ssl/example.test) which could be adapted to this section of the docs.

You've also missed removing the `docs-ssl` link at the bottom of the file as this is the only line referencing it, should this line remain removed entirely.

**Review by polarathene — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (wernerfred) [reply CHANGELOG.md:8]: > What does "more present" mean for documentation? Improved? Integrated? Visible? Or something else?

I think it is related to this PR: https://github.com/docker-mailserver/docker-mailserver/pull/1878 (reference can be added) I just copied the draft release in the beginning

Comment (polarathene) [reply CHANGELOG.md:8]:
```suggestion
- **[general]** Fixed many prose errors (spelling, grammar, indentation).
- **[general]** Documentation is better integrated into the development process and it's visibility within the project increased ([#1878](https://github.com/docker-mailserver/docker-mailserver/pull/1878)).
```

**Comment by aendeavor** [comment]
CI „fails“ because I am cancelling it as we have multiple commits that stand by themselves and this way, CI CPU time is not wasted (although we have unlimited, I‘m still uncomfortable :D). Just FYI.

File: docs/content/config/advanced/optional-config.md
Comment (aendeavor) [reply docs/content/config/advanced/optional-config.md:14]: I must have overlooked that. I will fix that.

**Review by aendeavor — APPROVED** [review]

**Review by polarathene — APPROVED** [review]
Note, as per [my prior review](https://github.com/docker-mailserver/docker-mailserver/pull/1978#discussion_r642299999), while the `setup.sh config ssl` command is being removed, it's still going to be [part of the v10 docs](https://docker-mailserver.github.io/docker-mailserver/edge/config/security/ssl/#self-signed-certificates).

Simple fix would be to just add an admonition notice about the command no longer being supported, or we could try revise a bit and suggest Bring Your Own Certs approach, perhaps mentioning [how we generate self-signed certs for tests](https://github.com/docker-mailserver/docker-mailserver/tree/master/test/test-files/ssl/example.test)?

**Comment by aendeavor** [comment]
> Note, as per [my prior review](https://github.com/docker-mailserver/docker-mailserver/pull/1978#discussion_r642299999), while the `setup.sh config ssl` command is being removed, it's still going to be [part of the v10 docs](https://docker-mailserver.github.io/docker-mailserver/edge/config/security/ssl/#self-signed-certificates).
> 
> Simple fix would be to just add an admonition notice about the command no longer being supported, or we could try revise a bit and suggest Bring Your Own Certs approach, perhaps mentioning [how we generate self-signed certs for tests](https://github.com/docker-mailserver/docker-mailserver/tree/master/test/test-files/ssl/example.test)?

Ah, now I understand properly. Can you briefly insert the „Bring Your Own Certs“ approach?

**Comment by casperklein** [comment]
> Had someone the chance to test a LDAP environment?

I don't have a LDAP environment. Beside that, all works as expected. I had no issues over the last weeks. Ready for release from my side.

**Review by casperklein — APPROVED** [review]

**Comment by polarathene** [comment]
> Can you briefly insert the „Bring Your Own Certs“ approach? I think this would be the best solution.

I'll update the section to advise how to generate a self-signed certificate like our tests do.

**EDIT:** At a glance over the TLS handling in `setup-stack.sh`, I'll keep the self-signed section to use `SSL_TYPE=self-signed` (which I have not written additional tests for) as there is handling for the self-signed CA required by the looks of it and I'm lacking time to contribute tests/refactor right now :sweat_smile: 

---

Do we want to drop `SSL_TYPE=self-signed` cert support completely in a later release?

The same certificates could use the custom SSL certificates section (_`SSL_TYPE=manual` or `SSL_TYPE=custom`_) right below it AFAIK, it's just a little bit more effort on the users part but I think that's fine (_no implicit filename and location, requires additional ENV_).

[Quick search of where `self-signed` is used in the project](https://github.com/docker-mailserver/docker-mailserver/search?q=self-signed).

**Comment by wernerfred** [comment]
So merging #2021 has to  happen before `v10`

**Comment by wernerfred** [comment]
Imho the `CHANGELOG.md` is now complete. If the build passes we are ready to go i think.
Awaiting confirmation/review from all of you again before merging this.

Anything missing?
@aendeavor who will publish the release afterwards? I'm going to copy the changelog in the `v10` release draft right now so after the merge and CI passing on master it should only be the press of a button

**Comment by aendeavor** [comment]
> @aendeavor who will publish the release afterwards? I'm going to copy the changelog in the `v10` release draft right now so after the merge and CI passing on master it should only be the press of a button

I leave you the honours :) I will then remove the `:stable` from DockerHub and GHCR.

**Review by aendeavor — APPROVED** [review]

**Review by polarathene — APPROVED** [review]

**Review by casperklein — APPROVED** [review]

**Comment by wernerfred** [comment]
Alright then, I'm merging this now and publish `v10` release after ci with `:edge` has successfully built on master

**Comment by wernerfred** [comment]
@polarathene we have an issue on v10 Docs deployment: https://github.com/docker-mailserver/docker-mailserver/runs/2717971173?check_suite_focus=true#step:4:9

Can we fix this / Push the version manually?

**Comment by polarathene** [comment]
:grimacing: whoops, that'd be my fault sorry!

We could modify the workflow to allow maintainers to trigger with `workflow_dispatch` and an input for tag/commit to deploy I think? Otherwise it should be fine to deploy manually to the `gh_pages` branch via PR/push afaik.

It's late here, but I can look into updating the workflow for this manual dispatch feature tomorrow if you'd like?

**Comment by wernerfred** [comment]
> It's late here, but I can look into updating the workflow for this manual dispatch feature tomorrow if you'd like?

No need to hurry. Take your time.

> Otherwise it should be fine to deploy manually to the gh_pages branch via PR/push afaik.

I will use this method for now to keep it streamlined just wanted your confirmation on this ;)

**Comment by wernerfred** [comment]
FYI @casperklein the update checker notified me right away the minute where i merged this PR:
image

However on this time the release was not yet published and no new image was pushed to dockerhub. Any idea on how to (easily) avoid such circumstances? No bid deal if not ;)

**Comment by casperklein** [comment]
That is the time frame between bumping the VERSION file and building the image. I don't think it's worth to fix that.. Default interval is 1d. So the chance to hit this little time window is low.

You could create a new branch. Bump the VERSION file there and build the new image from that branch. And after the build succeed, bump the VERSION also on master..

**Pull request merged**
